### PR TITLE
Update release packaging for Xcode 15/visionOS

### PR DESCRIPTION
**Pull request by tgoyne**
Most of this is cleanup of old legacy cruft that hasn't been needed for years.

Rather than a separate target, this switches to using a different configuration for the static framework build. This eliminates most of the duplication and hacks involved.

There's now just a single release package rather than separate objc and swift ones. As we no longer have to deal with the backwards-compatiblity issues from bitcode, only the obj-c framework built with the most recent version of Xcode is included rather than one per Xcode version as is done with Swift. The root directory contains the dynamic xcframework that most people should be using, and the static xcframework is in the `static` subdirectory. The static framework now could contain slices for all platforms rather than just iOS, but in practice that was a bit over 2 GB, so it continues to only include iOS to make the download a reasonable size.

RLMPlatform.h was made obsolete by XCFrameworks. Prior to those you had to manually link the correct Framework for your platform, and picking the wrong one resulted in linker errors that users didn't understand. Thankfully, that's no longer applicable.

The actual release publishing is now in a Jenkins declarative pipeline rather than being scattered over a bunch of different Jenkins jobs. A bunch of the incidental logic is now in build.sh rather than being tied specifically to Jenkins, which hopefully will help with migrating away from it.

REALM_SWIFT_VERSION was a legacy of the Swift 1/2/3 migration period where each version of the Swift compiler supported several different language versions and specifically selecting both the Xcode and Swift versions was required. Swift 4 moved away from that, and it now just complicated the logic.

Carthage has now had support for XCFrameworks for a few years and no one should be using the old fat frameworks any more, so stop publishing them.

Historically we didn't enable LTO because it made the build dramatically slower for tiny benefits, but the (relatively) new ThinLTO mostly eliminates the speed hit and meaningfully reduces the size of the binaries.

### CODE REVIEW

**Review by dianaafanador3 — COMMENTED** [review]
Pretty good job, moving release pipelines from the Jenkins file to `build.sh`. This will make the migration to Xcode cloud easier. 
Some minor comments and some questions

File: .jenkins.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ target:
  - swiftpm-thread
  - ios-xcode-spm
  - ios-static
- - ios-dynamic
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply .jenkins.yml:22]: It complicates the test build a bit to work with both static and dynamic linking, but it ended up being pretty minor.

File: Jenkinsfile.publish [hunk 2]
```diff
@@ -0,0 +1,177 @@
+def carthageXcodeVersion = '14.3.1'
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Jenkinsfile.publish:1]: It's a second pipeline which publishes the things built by the first one. This one is run by the new manually-triggered [Publish Cocoa Release](https://ci.realm.io/job/Publish%20Cocoa%20Release/) job which will replace the cocoa-release job.

File: Realm.xcodeproj/xcshareddata/xcschemes/Realm.xcscheme [hunk 3]
```diff
@@ -56,7 +56,7 @@
       </BuildActionEntries>
    </BuildAction>
    <TestAction
-      buildConfiguration = "Debug"
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Realm.xcodeproj/xcshareddata/xcschemes/Realm.xcscheme:59]: Accident that shouldn't be committed.

File: Configuration/Release.xcconfig [hunk 4]
```diff
@@ -8,4 +8,10 @@ DEBUG_INFORMATION_FORMAT = $(DEBUG_INFORMATION_FORMAT_$(XCODE_VERSION_MAJOR));
 
 ENABLE_NS_ASSERTIONS = NO;
 GCC_PREPROCESSOR_DEFINITIONS = REALM_HAVE_CONFIG REALM_ENABLE_SYNC __ASSERTMACROS__;
+LLVM_LTO = YES_THIN;
 VALIDATE_PRODUCT = YES;
+
+REALM_HIDE_SYMBOLS = NO;
+REALM_OTHER_LDFLAGS_NO = ;
+REALM_OTHER_LDFLAGS_YES = -Xlinker -unexported_symbols_list -Xlinker Configuration/Realm/PrivateSymbols.txt;
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply Configuration/Release.xcconfig:16]: It's just incredibly unwieldily to pass multiple exclusions on the command line. This currently doesn't add the flag to the cocoapods build, but if we add that then it's nice to not have to duplicate it for that.

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,8 @@ x.y.z Release notes (yyyy-MM-dd)
   release package. visionOS currently requires installing Realm via either
   Swift Package Manager or by using a XCFramework as CocoaPods and Carthage do
   not yet support it.
+* Zips compatible with SPM's `.binaryTarget()` are now published as part of the
```
[the remark this answers is not in the document]
Comment (dianaafanador3) [reply CHANGELOG.md:8]: https://github.com/realm/realm-swift/issues/6898

**Review by dianaafanador3 — APPROVED** [review]

**Pull request merged**